File: src-tauri/src/commands/gateway.rs
use crate::models::result::CommandResult;
use crate::services::gateway_service::{
    self, DashboardProbeData, GatewayActionData, GatewayStatusData,
};

#[tauri::command]
pub async fn get_gateway_status() -> CommandResult<GatewayStatusData> {
    match gateway_service::get_gateway_status().await {
        Ok(data) => CommandResult::ok(data),
        Err(error) => CommandResult::err(error),
    }
}

#[tauri::command]
pub async fn start_gateway() -> CommandResult<GatewayActionData> {
    match gateway_service::start_gateway().await {
        Ok(data) => CommandResult::ok(data),
        Err(error) => CommandResult::err(error),
    }
}

#[tauri::command]
pub async fn stop_gateway() -> CommandResult<GatewayActionData> {
    match gateway_service::stop_gateway().await {
        Ok(data) => CommandResult::ok(data),
        Err(error) => CommandResult::err(error),
    }
}

#[tauri::command]
pub async fn restart_gateway() -> CommandResult<GatewayActionData> {
    match gateway_service::restart_gateway().await {
        Ok(data) => CommandResult::ok(data),
        Err(error) => CommandResult::err(error),
    }
}

#[tauri::command]
pub async fn open_dashboard() -> CommandResult<GatewayActionData> {
    match gateway_service::open_dashboard().await {
        Ok(data) => CommandResult::ok(data),
        Err(error) => CommandResult::err(error),
    }
}

#[tauri::command]
pub async fn probe_dashboard_endpoint(address: String) -> CommandResult<DashboardProbeData> {
    match gateway_service::probe_dashboard_endpoint(address).await {
        Ok(data) => CommandResult::ok(data),
        Err(error) => CommandResult::err(error),
    }
}

File: src/main/ipc/gateway.ts
import { ipcMain } from "electron";
import { IPC } from "../../shared/ipc-channels.js";
import { ok, err } from "../models/result.js";
import { toAppError } from "./utils.js";
import {
  getGatewayStatus,
  startGateway,
  stopGateway,
  restartGateway,
  openDashboard,
  probeDashboardEndpoint,
} from "../services/gateway-service.js";

export function registerGatewayHandlers(): void {
  ipcMain.handle(IPC.GET_GATEWAY_STATUS, async () => {
    try { return ok(await getGatewayStatus()); }
    catch (error) { return err(toAppError(error)); }
  });

  ipcMain.handle(IPC.START_GATEWAY, async () => {
    try { return ok(await startGateway()); }
    catch (error) { return err(toAppError(error)); }
  });

  ipcMain.handle(IPC.STOP_GATEWAY, async () => {
    try { return ok(await stopGateway()); }
    catch (error) { return err(toAppError(error)); }
  });

  ipcMain.handle(IPC.RESTART_GATEWAY, async () => {
    try { return ok(await restartGateway()); }
    catch (error) { return err(toAppError(error)); }
  });

  ipcMain.handle(IPC.OPEN_DASHBOARD, async () => {
    try { return ok(await openDashboard()); }
    catch (error) { return err(toAppError(error)); }
  });

  ipcMain.handle(IPC.PROBE_DASHBOARD_ENDPOINT, async (_event, payload: { address: string }) => {
    try { return ok(await probeDashboardEndpoint(payload.address)); }
    catch (error) { return err(toAppError(error)); }
  });
}
